executor/insert_common.go

```diff
@@ -455,33 +457,126 @@ func (e *InsertValues) setValueForRefColumn(row []types.Datum, hasValue []bool)
 	return nil
 }
 
-func insertRowsFromSelect(ctx context.Context, base insertCommon) error {
+func insertRowsFromSelect(ctx context.Context, base insertCommon, etlMode bool) (err error) {
 	// process `insert|replace into ... select ... from ...`
 	e := base.insertCommon()
 	selectExec := e.children[0]
 	fields := retTypes(selectExec)
 	chk := tryNewCacheChunk(selectExec)
 	iter := chunk.NewIterator4Chunk(chk)
 	rows := make([][]types.Datum, 0, chk.Capacity())
+	var chunkPool *sync.Pool
 
 	sessVars := e.ctx.GetSessionVars()
 	batchSize := sessVars.DMLBatchSize
 	batchInsert := sessVars.BatchInsert && !sessVars.InTxn() && variable.EnableBatchDML.Load() && batchSize > 0
 	memUsageOfRows := int64(0)
 	memUsageOfExtraCols := int64(0)
 	memTracker := e.memTracker
+
+	var (
+		dataCh chan *chunk.Chunk
+	)
+	etlConcurrency, etlBatchSize := sessVars.ETLConcurrency, sessVars.ETLBatchSize
+	if etlConcurrency <= 1 {
+		etlMode = false
+	}
+	if etlMode {
+		var workers []insertCommon
+		workers, err = createFromSelectWorkers(ctx, base, etlConcurrency)
+		if err != nil {
+			return err
+		} else if len(workers) == 0 {
+			// executor not support, fallback to single thread.
+			etlMode = false
+		} else {
+			affectedRow := uint64(0)
+			maxCap := e.maxChunkSize
+			approximateChannelSize := 2 * etlConcurrency * (batchSize/maxCap + 1)
```

Comment:
`2*etlConcurrency` is enough. A worker will fetch the next chunk if it wants `etlBatchSize` enough rows.

Reply:
`2 * etlConcurrency * (batchSize/maxCap + 1)` is a little better in my test, guess it's because calling `Next` may make the worker threads wait.